### PR TITLE
perf: add threading benchmark scenario

### PR DESCRIPTION
**Pull request by brettlangdon**
This benchmark test is used to simulate the creation, encoding, and flushing of traces in threaded environments.

It uses a ``concurrent.futures.ThreadPool`` to manage the total number of workers.

The only modification to the tracing workflow that has been modified is the ``AgentWriter``'s ``_send_payload`` has been overridden to do nothing.
This means we will still start the background worker for the agent, put traces into the queue, and encode them into payloads to send. We will just
not try to make the HTTP call to the agent to reduce any impact from an external process on this test.

### CODE REVIEW

**Comment by brettlangdon** [comment]
I marked as manual because locally it won't show me all results because they are not significant, I want to be sure running this on dedicated performance testing hardware shows it as being significant, otherwise it won't be a useful scenario to have.


```
+--------------------+---------------------+----------------------+
| Benchmark          | 0.50.4/results.json | 0.51.2/results.json  |
+====================+=====================+======================+
| threading-1-thread | 594 ms              | 608 ms: 1.02x slower |
+--------------------+---------------------+----------------------+
| Geometric mean     | (ref)               | 1.00x faster         |
+--------------------+---------------------+----------------------+

Benchmark hidden because not significant (3): threading-10-threads, threading-50-threads, threading-100-threads
```

**Comment by brettlangdon** [comment]
Discussed with @majorgreys off of GitHub. Although pyperf has issues with these tests because of significance, they might still be useful for us. We might want to consider going with the same strategy that the `benchmarks/tracer` scenario uses which is to have a filter that just drops traces to skip the encoding/queuing/flushing process. The periodic flushing is probably what is messing up these tests and unless we have a better way to control it (we did try and explicit tracer shutdown at the end of every run) then we might be better off cutting the scope of what this is testing.


This means we will target trace creation + context management in a threaded application.

**Review by majorgreys — COMMENTED** [review]
Can we include these docs in our benchmarking section? We could add docs for the other scenarios as well.

**Comment by brettlangdon** [comment]
> Can we include these docs in our benchmarking section? We could add docs for the other scenarios as well.

@majorgreys I updated to include in the main `benchmarks/README.rst`, is this what you had in mind?

image

**Review by majorgreys — DISMISSED** [review]

File: benchmarks/threading/README.rst [hunk 1]
```diff
@@ -0,0 +1,10 @@
+threading
+~~~~~~~~~
+
+This benchmark test is used to simulate the creation, encoding, and flushing of traces in threaded environments.
+
+It uses a ``concurrent.futures.ThreadPool`` to manage the total number of workers.
+
+The only modification to the tracing workflow that has been modified is using a ``NoopWriter`` which does not start a
```
Comment (P403n1x87) [inline benchmarks/threading/README.rst:8]:
```suggestion
The only modification to the tracing workflow that has been made is using a ``NoopWriter`` which does not start a
```

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Pull request merged**